core/src/cli_commands/update_pool.rs

```diff
@@ -0,0 +1,629 @@
+    wallet::WalletDataSource,
+};
+use derive_more::From;
+use thiserror::Error;
+
+#[derive(Debug, Error, From)]
+pub enum UpdatePoolError {
+    #[error("Update pool: Not enough votes, expected {0}, found {1}")]
+    NotEnoughVotes(usize, usize),
+    #[error("Update pool: Pool parameters (refresh NFT, update NFT) unchanged")]
+    PoolUnchanged,
+    #[error("Update pool: ErgoBoxCandidateBuilderError {0}")]
+    ErgoBoxCandidateBuilder(ErgoBoxCandidateBuilderError),
+    #[error("Update pool: box selector error {0}")]
+    BoxSelector(BoxSelectorError),
+    #[error("Update pool: tx builder error {0}")]
+    TxBuilder(TxBuilderError),
+    #[error("Update pool: tx context error {0}")]
+    TxSigningError(TxSigningError),
+    #[error("Update pool: stage error {0}")]
+    StageError(StageError),
+    #[error("Update pool: node error {0}")]
+    Node(NodeError),
+    #[error("No change address in node")]
+    NoChangeAddressSetInNode,
+    #[error("Update pool: address encoder error {0}")]
+    AddressEncoderError(AddressEncoderError),
+    #[error("Update pool: pool contract error {0}")]
+    PoolContractError(crate::contracts::pool::PoolContractError),
+    #[error("Update pool: io error {0}")]
+    IoError(std::io::Error),
+    #[error("Update pool: yaml error {0}")]
+    YamlError(serde_yaml::Error),
+    #[error("Update pool: could not find unspent wallot boxes that do not contain ballot tokens")]
+    NoUsableWalletBoxes,
+}
+
+pub fn update_pool(
+    new_pool_box_hash_str: Option<String>,
+    reward_token_id: Option<String>,
+    reward_token_amount: Option<u64>,
+    height: Option<u64>,
+) -> Result<(), UpdatePoolError> {
+    info!("Opening oracle_config_updated.yaml");
+    let s = std::fs::read_to_string("oracle_config_updated.yaml")?;
+    let new_oracle_config: OracleConfig = serde_yaml::from_str(&s)?;
+    let wallet = crate::wallet::WalletData {};
+    let op = OraclePool::new().unwrap();
+    let change_address_str = get_wallet_status()?
+        .change_address
+        .ok_or(UpdatePoolError::NoChangeAddressSetInNode)?;
+
+    let network_prefix = if ORACLE_CONFIG.on_mainnet {
+        NetworkPrefix::Mainnet
+    } else {
+        NetworkPrefix::Testnet
+    };
+    let change_address =
+        AddressEncoder::new(network_prefix).parse_address_from_str(&change_address_str)?;
+
+    let pool_contract_inputs = PoolContractInputs::from((
+        &new_oracle_config.pool_contract_parameters,
+        &new_oracle_config.token_ids,
+    ));
+
+    let new_pool_contract = PoolContract::new(pool_contract_inputs)?;
+    let new_pool_box_hash = blake2b256_hash(
+        &new_pool_contract
+            .ergo_tree()
+            .sigma_serialize_bytes()
+            .unwrap(),
+    );
+
+    display_update_diff(
+        &ORACLE_CONFIG,
+        &new_oracle_config,
+        op.get_pool_box_source().get_pool_box()?,
+        None,
+    );
+
+    if new_pool_box_hash_str.is_none() {
```

Comment:
The hash string/height is only to prevent mistakes as per the original issue. Currently if you run it without any arguments, it will display the update diff including the new pool box hash, and will only build/submit the update transaction once you pass the right parameters to it

Reply:
Hmm, the height in R5 of ballots is actually different from current block height. It's to show the creation height of the update NFT/box.

Reply:
You're right. I rechecked the code and the EIP, and I cannot find a reason why I added the height parameter back then. I just realized that I wrote my previous comment and started this comment in the wrong "thread" :), it was meant to be for the above optional height parameter. I understand now the reason for pool box hash being optional. Good call!
As for the height, I suspect I added it initially because I messed up R5 meaning in the ballot box, so it's safe to remove it. What do you think?

Reply:
Agreed.